helm/revoke.go

```diff
@@ -0,0 +1,155 @@
+package helm
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/gruntwork-io/kubergrunt/kubectl"
+	"github.com/gruntwork-io/kubergrunt/logging"
+)
+
+// RevokeAccess revokes access to a Tiller pod from a provided RBAC user, group, or serviceaccount in a
+// provided Tiller namespace by deleting the secret, role, and rolebindings associated with said entities.
+// Note that due to limitations in the Go TLS library used by helm, helm/tiller does not support checking certificate revocation lists. As a
+// consequence, the signed TLS certificate will continue to be trusted by Tiller after running "kubergrunt helm revoke" since
+// it was signed by the Tiller CA. However, the user's authorizations are removed by way of deleting the role and role binding and access is
+// effectively removed.
+// See https://github.com/helm/helm/issues/4273
+func RevokeAccess(
+	kubectlOptions *kubectl.KubectlOptions,
+	tillerNamespace string,
+	rbacGroups []string,
+	rbacUsers []string,
+	serviceAccounts []string,
+) error {
+	logger := logging.GetProjectLogger()
+	logger.Infof(
+		"Revoking access to Tiller server deployed in namespace %s to:",
+		tillerNamespace,
+	)
+	logEntities(rbacGroups, rbacUsers, serviceAccounts)
+
+	logger.Info("Checking if Tiller is deployed in the namespace.")
+	if err := validateTillerDeployed(kubectlOptions, tillerNamespace); err != nil {
+		logger.Errorf("Did not find a deployed Tiller instance in the namespace %s.", tillerNamespace)
+		return err
+	}
+	logger.Infof("Found a valid Tiller instance in the namespace %s.", tillerNamespace)
+
+	if len(rbacGroups) > 0 {
+		logger.Infof("Revoking access from RBAC groups")
+		if err := revokeAccessFromRBACEntities(kubectlOptions, tillerNamespace, convertToGroupInfos(rbacGroups)); err != nil {
+			return err
+		}
+		logger.Infof("Successfully revoked access from RBAC groups")
+	} else {
+		logger.Infof("No RBAC groups to revoke - skipping")
+	}
+
+	if len(rbacUsers) > 0 {
+		logger.Infof("Revoking access from RBAC users")
+		if err := revokeAccessFromRBACEntities(kubectlOptions, tillerNamespace, convertToGroupInfos(rbacUsers)); err != nil {
+			return err
+		}
+		logger.Infof("Successfully revoked access from RBAC users")
+	} else {
+		logger.Infof("No RBAC users to revoke - skipping")
+	}
+
+	if len(serviceAccounts) > 0 {
+		logger.Infof("Revoking access from RBAC Service Accounts")
+		serviceAccountInfos, err := convertToServiceAccountInfos(serviceAccounts)
+		if err != nil {
+			return err
```

Comment:
If we return an error part way through, is there any chance of us ending up in some weird partial state that can't be recovered from? That is, the user's access is partially revoked, but not fully, and some possible error condition as a result?